### PR TITLE
Array implementation

### PR DESCRIPTION
**Pull request by maasencioh**
Work on progress

### CODE REVIEW

**Comment by maasencioh** [comment]
@mljs/collaborators this is only the array implementation, the idea is to extend it to matrix but for now I prefer to merge it now and see the rest later, I'm even thinking to publish in npm in order to use it for Savitsky-Golay and other ML packages

**Comment by targos** [comment]
Can you put the whole output array in the tests ?
Something like:
```js
it('Default test', function () {
        var data = [1, 2, 3, 4];
        var model = padArray(data);
        model.should.eql([0, 1, 2, 3, 4, 0]);
});
```

File: README.md [hunk 1]
```diff
@@ -1 +1,73 @@
...
+
+This code is based in the MATLAB's code [padarray](http://www.mathworks.com/help/images/ref/padarray.html)
+
+## Installation
+
+`npm install ml-pad-array`
+
+## pad-array(data, [options])
+
+Pads the array `data` given the current options`.
```
Comment (targos) [inline README.md]: unwanted backtick after options

File: README.md [hunk 2]
```diff
@@ -1 +1,73 @@
 # pad-array
+
+[![NPM version][npm-image]][npm-url]
+[![build status][travis-image]][travis-url]
+[![David deps][david-image]][david-url]
```
Comment (targos) [inline README.md]: you can remove the david badge if the project doesn't have dependencies

File: README.md [hunk 3]
```diff
@@ -1 +1,73 @@
...
+## License
+
+[MIT](./LICENSE)
+
+[npm-image]: https://img.shields.io/npm/v/ml-pad-array.svg?style=flat-square
+[npm-url]: https://npmjs.org/package/ml-pad-array
+[travis-image]: https://img.shields.io/travis/mljs/pad-array/master.svg?style=flat-square
+[travis-url]: https://travis-ci.org/mljs/pad-array
+[david-image]: https://img.shields.io/david/mljs/pad-array.svg?style=flat-square
+[david-url]: https://david-dm.org/mljs/pad-array
```
Comment (targos) [inline README.md]: both lines can be removed

File: package.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  },
+  "keywords": [
+    "pad",
+    "fill",
+    "array",
+    "utils"
+  ],
+  "author": {
+    "name": "Miguel Asencio",
+    "mail": "maasencioh@gmail.com"
```
Comment (targos) [inline package.json]: I think this field is supposed to be "email"

File: README.md (hunk 1 above)
Comment (targos) [inline README.md]: may be good to specify that it doesn't change the input array and returns a new one

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+ * Pads and array
+ * @param {Array <number>} data
+ * @param {object} options
+ */
+function padArray (data, options) {
+    options = options || {};
+    options.padsize = options.padsize || defaultOptions.padsize;
+    options.padval  = options.padval  || defaultOptions.padval;
+
+    if (data[0] === undefined)
```
Comment (targos) [inline src/index.js]: the fact that `data[0]` exists is not proof that it is an array. I'd use `Array.isArray(data)`

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        throw new TypeError('data should be an array');
+    else if (data[0][0] === undefined)
```
Comment (targos) [inline src/index.js]: ditto, `Array.isArray(data[0][0])`

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+function arrayCase(data, options) {
+    var len = data.length;
+    if (options.padsize[0] === undefined)
+        options.padsize = [options.padsize, options.padsize];
+
+    var ans = new Array(len + options.padsize[0] + options.padsize[1]);
+
+    // circular option
+    if (options.padval === 'circular') {
+        for (var i = 0; i < (len + options.padsize[0] + options.padsize[1]); i++) {
```
Comment (targos) [inline src/index.js]: I'm not sure the VM is clever enough to optimize the condition.

Comment (targos) [inline src/index.js]: it is not: http://jsperf.com/loop-access-array-in-condition

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            else if (k < (options.padsize[0] + len))
+                ans[k] = data[k - options.padsize[0]];
+            else
+                ans[k] = data[2*len + options.padsize[0] - k - 1];
+        }
+    }
+
+    // default option
+    else {
+        for (var l = 0; l < (len + options.padsize[0] + options.padsize[1]); l++) {
```
Comment (targos) [inline src/index.js]: not very important but you don't need to declare a different variable for each case (they will be all hoisted to the beginning of the function).
You can simply do `var i;` at the beginning and start the loop with `for(i = 0; ...)`

**Comment by targos** [comment]
made a few comments but looks good !

**Comment by targos** [comment]
:+1: thank you!

**Comment by maasencioh** [comment]
no man thanks to you, the question is should I publish it now to NPM and put it in ML or should I wait until I finish the matrix part?

**Comment by lpatiny** [comment]
I think it should be published without waiting for Matrix

**Comment by targos** [comment]
I agree

**Pull request merged**